pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -968,15 +964,18 @@ func (cm *containerManagerImpl) GetDynamicResources(pod *v1.Pod, container *v1.C
 		// a set of CDIDevices from a different kubelet plugin. In the future we may want to
 		// include the name of the kubelet plugin and/or other types of resources that are
 		// not CDIDevices (assuming the DRAmanager supports this).
-		for _, klPluginCdiDevices := range containerClaimInfo.CDIDevices {
+		for /* driverName */ _, driverState := range containerClaimInfo.Drivers {
 			var cdiDevices []*podresourcesapi.CDIDevice
-			for _, cdiDevice := range klPluginCdiDevices {
-				cdiDevices = append(cdiDevices, &podresourcesapi.CDIDevice{Name: cdiDevice})
+			for /*requestName */ _, requestDevices := range driverState.CDIDevices {
+				for _, cdiDevice := range requestDevices {
+					cdiDevices = append(cdiDevices, &podresourcesapi.CDIDevice{Name: cdiDevice})
+				}
 			}
 			claimResources = append(claimResources, &podresourcesapi.ClaimResource{CDIDevices: cdiDevices})
 		}
 		containerDynamicResource := podresourcesapi.DynamicResource{
-			ClassName:      containerClaimInfo.ClassName,
+			// As of 1.31, ResourceClaims are not associated with a single class anymore.
+			// ClassName: ???
```

Comment:
Given the purpose of classes now, I think this is OK to drop.

Reply:
That said, we will now need to store info about the driver used to allocate each resource though. Not sure of the best way to represent that off the top of my head.

Reply:
Claims are also not associated with a single driver. I suspect that whatever this `podresourcesapi.DynamicResource` API is, it made some premature assumptions about how DRA works.

We need to be a bit more careful about what we expose here.

Reply:
As far as I know, the only consumber of it at the moment is CNI plugins that need to see what resources have been allocated so they can set up networks appropriately.

Reply:
Looks like ClassName should be pushed down into `podresourcesapi.ClaimResource`, which currently just contains a list of CDIDevices. Or should that be changed to be `ClaimDeviceRequest`, which then corresponds back to our `DeviceRequest` API. This would mirror the current control plane side structure.

Reply:
What's unclear to me though is why they need to know the groupings and hierarchy used on the control plane during the allocation process. Are we better off minimizing the information to exactly what is needed, at least until we have a beta?

Reply:
This is where they use it:
https://github.com/k8snetworkplumbingwg/multus-cni/blob/41013e7580396629a02213019baad1d0c84a6a2c/pkg/kubeletclient/kubeletclient.go#L175

Reply:
Yeah, this doesn't seem right with our new structure. If I understand [this code](https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/pkg/k8sclient/k8sclient.go#L275) correctly (which I might not because there is a lot there), these class names are referenced as if they were resource names in a multus annotation. The devices are bucketed by class and then associated with a CNI config. That's not the right way to do this, I don't think. Instead, we should probably use Config resources to make the association.

@moshe010 maybe we can discuss how to integrate with multus in the 1.31 way? cc @aojea

Reply:
See also https://github.com/kubernetes/kubernetes/pull/125488/commits/eefc9730ebf03fc3adbff006048851a94460b4c4#r1665538721

Reply:
k8s.io/kubelet/pkg/apis/podresources/v1 is a GA API of kubelet. I suppose all aspects of it related to DRA are considered alpha and can be changed?

Here's a potential change to that API:
```patch
diff --git a/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto b/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto
index 9e6022ac32c..5975fc0a6f6 100644
--- a/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto
+++ b/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto
@@ -82,14 +82,20 @@ message NUMANode {
 
 // DynamicResource contains information about the devices assigned to a container by DRA
 message DynamicResource {
-    string class_name = 1;
+    // tombstone: removed in 1.31 because claims are no longer associated with one class
+    // string class_name = 1;
     string claim_name = 2;
     string claim_namespace = 3;
     repeated ClaimResource claim_resources = 4;
 }
 
-// ClaimResource contains per plugin resource information
+// ClaimResource contains resource information. The driver name/pool name/device name
+// triplet uniquely identifies the device. Should DRA get extended to other kinds
+// of resources, then device_name will be empty and other fields will get added.
 message ClaimResource {
+    string driver_name = 1;
+    string pool_name = 2;
+    string device_name = 3;
     repeated CDIDevice cdi_devices = 1 [(gogoproto.customname) = "CDIDevices"];
 }
```

Reply:
Yes, the DRA aspects are considered alpha. They only get populated if the DRA feature gate is enabled. The change you are proposing seems reasonable. I think multus just used the claim name as a means to an end, and using `<driver_name>/<pool_name>` should be just as good (possibly even better since `pool_name` makes it very specific to this node now).

/cc @aojea @adrianchiris @moshe010 
for their thoughts

Reply:
To implement this, I had to extend the NodePublishResources API. Otherwise kubelet wouldn't have known which CDI devices belong to with device at the API level. I'll comment on the source because this thread here got detached from source code.